src/devices/bus/ieee488/hp9133.cpp

```diff
@@ -0,0 +1,878 @@
+		break;
+	}
+
+	if (!m_intenable) {
+		irq = false;
+		firq = false;
+	}
+
+	if (m_cpuirq != irq) {
+		m_cpu->set_input_line(M6809_IRQ_LINE, irq ? ASSERT_LINE : CLEAR_LINE);
+		m_cpuirq = irq;
+	}
+
+	if (m_cpufirq != firq) {
+		m_cpu->set_input_line(M6809_FIRQ_LINE, firq ? ASSERT_LINE : CLEAR_LINE);
+		m_cpufirq = firq;
+	}
+}
+
+TIMER_CALLBACK_MEMBER(hp9133_device::fast_timer)
+{
+	if (!m_fast)
+		return;
+	if (m_gpib_drq) {
+		if (!m_dma_rwn)
+			m_gpib->dout_w(dma_ram_r(0));
+		else
+			dma_ram_w(m_gpib->din_r());
+	}
+	m_fast_timer->adjust(attotime::from_usec(50));
+}
+
+void hp9133_device::io1_w(uint8_t data)
+{
+	// 7 - HDC Rst#
+	// 6 - HDC BRDY
+	// 5 - HD Precomp#
+	// 4 - HD Head 2
+	// 3 - HD Head 1
+	// 2 - HD Head 0
+	// 1 - Active LED
+	// 0 - Fault LED#
+	m_head = (data >> 2) & 7;
+	LOG("%s: %02x Head %d %s%s%s%s\n", __func__,
+	    data, m_head,
+	    data & IO1_W_BRDY ? "BRDY " : "",
+	    data & IO1_W_WDRESET_N ? "" : "HDCRESET ",
+	    data & IO1_W_DS1 ? "ACTIVE " : "",
+	    data & IO1_W_FLT_N ? "" : "FAULT ");
+	if (!(data & IO1_W_WDRESET_N))
+		m_hdc->reset();
+	m_hdc->buffer_ready(data & IO1_W_BRDY);
+}
+
+void hp9133_device::io2_w(uint8_t data)
+{
+	auto *floppy = m_floppy->get_device();
+
+	// 7 - HDRecMode
+	// 6 - FlopHLO#?
+	// 5 - FlopMotor#
+	// 4 - FlopSide#
+	// 3 - Floppy Disk change reset
+	// 2 - ClrIdxInt
+	// 1
+	// 0 - FDCRst#
+	LOG("%s: %02x\n", __func__, data);
+	if (!(data & IO2_W_CLRIND_N))
+		m_floppy_index_int = false;
+	if (data & IO2_W_FDMOTON_N) {
+		m_motor_timer->adjust(attotime::from_msec(2000));
+	} else {
+		m_motor_timer->reset();
+		floppy->mon_w(0);
+	}
+	if (!(data & IO2_W_DKCHRES_N))
+		m_floppy->get_device()->dskchg_w(false);
+	m_floppy->get_device()->ss_w((data & IO2_W_FDHDSEL_N) ? 0 : 1);
+	if (!(data & IO2_W_FDCRES_N))
+		m_fdc->reset();
+}
+
+void hp9133_device::io3_w(uint8_t data)
+{
+	// 7 - DMA enable
+	// 6 - 1 GPIPACK, 0 FDCACK
+	// 5 - Write Precomp?
+	// 4 - FAST enable?
+	// 3 -
+	// 2 - Interrupt Enable
+	// 0:1 - Interrupt select
+
+	m_dmaenable = data & IO3_W_DMA;
+	m_dmaack_switch = data & IO3_W_HPB_FDC_N;
+	m_dma_rwn = data & IO3_W_DMAR_W_N;
+	m_intenable = !(data & IO3_W_INTENBL_N);
+	m_fast = data & IO3_W_FAST;
+	m_intsel = data & IO3_W_INTSEL_MASK;
+
+	LOG("%s: %02x = INTSEL %d, INT %d DMA %d DMA start %d DMA ACK %s\n", __func__,
+	    data, m_intsel, m_intenable, m_dmaenable, (data >> 4) & 1,
+	    m_dmaack_switch ? "GPIB" : "FDC");
+	update_intsel();
+	if (m_fast)
+		m_fast_timer->adjust(attotime::from_usec(1000));
+	else
+		m_fast_timer->adjust(attotime::never);
+}
+
+uint8_t hp9133_device::io1_r(void)
+{
+	uint8_t ret;
+
+	// 7 - hdindex
+	// 6 - HDC int
+	// 5 - HDC DRQ
+	// 4 - HDC BCS
+	// 3 - 0 Ident
+	ret = m_model->read() ^ 0xf;
+	if (m_harddisk_index)
+		ret |= IO1_R_HDIND;
+	if (m_hdc_intrq)
+		ret |= IO1_R_HDINT;
+	if (m_hdc_bdrq)
+		ret |= IO1_R_BDRQ;
+	if (m_hdc_bcs)
+		ret |= IO1_R_BCS_N;
+	LOG("%s: %02x\n", __func__, ret);
+	return ret;
+}
```

Comment:
It’s generally polite to suppress log messages when side effects are disabled to avoid spamming the log when a debugger window is open.

Reply:
Ok.